athenzdomains.go

```diff
@@ -0,0 +1,183 @@
+	crMap := domainMap[domainName]
+	for _, role := range item.Spec.SignedDomain.Domain.Roles {
+		crMap.RoleToPrincipals[string(role.Name)] = role.RoleMembers
+	}
+
+	for _, policy := range item.Spec.SignedDomain.Domain.Policies.Contents.Policies {
+		for _, assertion := range policy.Assertions {
+			_, ok := crMap.RoleToAssertion[assertion.Role]
+			if !ok {
+				crMap.RoleToAssertion[assertion.Role] = []SimpleAssertion{}
+			}
+			effect := assertion.Effect.String()
+			resourceRegex, err := regexp.Compile("^" + replacer.Replace(strings.ToLower(assertion.Resource)) + "$")
+			if err != nil {
+				fmt.Printf("Error occurred when converting assertion resource into regex format. Error: %v", err)
+			}
+			actionRegex, err := regexp.Compile("^" + replacer.Replace(strings.ToLower(assertion.Action)) + "$")
+			if err != nil {
+				fmt.Printf("Error occurred when converting assertion action into regex format. Error: %v", err)
+			}
+			simpleAssert := SimpleAssertion{
+				resource: resourceRegex,
+				action:   actionRegex,
+				effect:   effect,
+			}
+			crMap.RoleToAssertion[assertion.Role] = append(crMap.RoleToAssertion[assertion.Role], simpleAssert)
+		}
+
+	}
+}
+
+func (c *Cache) addObj(domainMap map[string]CRMap, item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := domainMap[domainName]
+	if !ok {
+		roleToPrincipals := make(map[string][]*zms.RoleMember)
+		roleToAssertion := make(map[string][]SimpleAssertion)
+		crMap := CRMap{
+			RoleToPrincipals: roleToPrincipals,
+			RoleToAssertion:  roleToAssertion,
+		}
+		domainMap[domainName] = crMap
+	}
+	parseData(domainMap, domainName, item)
+	c.lock.Unlock()
+}
+
+func (c *Cache) updateObj(domainMap map[string]CRMap, item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := domainMap[domainName]
+	if ok {
+		delete(domainMap, domainName)
+	}
+	roleToPrincipals := make(map[string][]*zms.RoleMember)
+	roleToAssertion := make(map[string][]SimpleAssertion)
+	crMap := CRMap{
+		RoleToPrincipals: roleToPrincipals,
+		RoleToAssertion:  roleToAssertion,
+	}
+	domainMap[domainName] = crMap
+	parseData(domainMap, domainName, item)
+	c.lock.Unlock()
+}
+
+func (c *Cache) deleteObj(domainMap map[string]CRMap, item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := domainMap[domainName]
+	if ok {
+		delete(domainMap, domainName)
+	}
+	c.lock.Unlock()
+}
+
+func (c *Cache) authorize(ctx context.Context, principal string, check AthenzAccessCheck) {
+	domainName := strings.Split(check.Resource, ":")
+	if len(domainName) != 2 {
+		fmt.Println("Error splitting domain name")
+	}
+
+	c.lock.RLock()
+	crMap := c.DomainMap
+	domainData := crMap[domainName[0]]
+	roles := []string{}
+	for roleName, member := range domainData.RoleToPrincipals {
+		for _, m := range member {
+			memberRegex, err := regexp.Compile("^" + replacer.Replace(strings.ToLower(string(m.MemberName))) + "$")
```

Comment:
Can we move the regex compilation to the parseData function so it's done offline? This is expensive to do, especially for every single call.